File: java-solutions/src/google/MergeSortedLists.java
package google;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

// merge two sorted lists into a new list
public class MergeSortedLists {

    // merge lists into a new list, and return it
    public List<Integer> mergeLists(List<Integer> ls, List<Integer> ms) {
        ArrayList<Integer> merged = new ArrayList<>();
        int lsi = 0;
        int msi = 0;
        while (lsi < ls.size() && msi < ms.size()) {
            if (ls.get(lsi) < ms.get(msi)) {
                merged.add(ls.get(lsi));
                lsi += 1;
            } else {
                merged.add(ms.get(msi));
                msi += 1;
            }
        }
        // if ls still has elements in it, add them to the merged list
        if (lsi < ls.size()) {
            while (lsi < ls.size()) {
                merged.add(ls.get(lsi++));
            }
        } else {
            // ms still has elements in it, add them to merged list
            while (msi < ms.size()) {
                merged.add(ms.get(msi++));
            }
        }
        return merged;
    }

    // combine ms into ls, mutating ls
    public void combineLists(List<Integer> ls, List<Integer> ms) {
        int lsi = 0;
        int msi = 0;
        while (lsi < ls.size() && msi < ms.size()) {
            if (ms.get(msi) < ls.get(lsi)) {
                ls.add(lsi, ms.get(msi));
                msi += 1;
            } else {
                lsi += 1;
            }
        }
        // if ms still has elements in it, add them to the merged list
        if (msi < ms.size()) {
            while (msi < ms.size()) {
                ls.add(ms.get(msi++));
            }
        }
    }

    public static void main(String[] args) {
        MergeSortedLists msl = new MergeSortedLists();
        List<Integer> l1 = new ArrayList<>(Arrays.asList(2,4,6,8,10));
        List<Integer> l2 = new ArrayList<>(Arrays.asList(1,3,5));
        msl.combineLists(l2, l1);
        System.out.println(l2);
    }
}

File: rust-solutions/google/src/bin/merge-sorted-lists.rs

/// merge two sorted vectors into a new sorted vector
fn merge_lists<T: Ord + Copy>(ls: &Vec<T>, ms: &Vec<T>) -> Vec<T> {

    let mut ts: Vec<T> = Vec::new();
    let mut li = 0;
    let mut mi = 0;

    while li < ls.len() && mi < ms.len() {
        if ls[li] <= ms[mi] {
            ts.push(ls[li]);
            li += 1;
        } else {
            ts.push(ms[mi]);
            mi += 1;
        }
    }

    while li < ls.len() {
        ts.push(ls[li]);
        li += 1;
    }

    while mi < ms.len() {
        ts.push(ms[mi]);
        mi += 1;
    }

    ts
}


fn main() {
    let l1 = vec![2,4,6];
    let l2 = vec![1,3,5,7,9];
    let l3 = vec![10, 20, 30, 40, 50, 60, 70];
    let l4 = vec![11,22,33,44,55];


    dbg!(merge_lists(&l1, &l2));
    dbg!(merge_lists(&l3,&l4));

}